### PR TITLE
Add Flex Sync, update to Core 12.5.1

### PR DESCRIPTION
**Pull request by leemaguire**
This PR adds the following:

- Upgrade to Core 12.5.1
- Add support for bool properties and lists
- Add ability to access an object via index subscript on results.
- Add support for Flexible Sync

### CODE REVIEW

File: src/cpprealm/persisted.hpp [hunk 1]
```diff
@@ -778,6 +774,10 @@ class rbool {
     operator bool() {
         return b;
     }
+    union {
```
Comment (jsflax) [inline src/cpprealm/persisted.hpp:777]: This should not be exposed publicly.

File: src/cpprealm/flex_sync.hpp [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+    SharedRealm m_realm;
+
+    friend struct SyncSubscriptionSet;
+    sync::MutableSubscriptionSet get_subscription_set()
+    {
+        return m_subscription_set;
+    }
+};
+
+struct SyncSubscriptionSet {
```
Comment (jsflax) [inline src/cpprealm/flex_sync.hpp:129]: Should be lower snake case. Though this isn't really cpp naming convention, we currently need the layer of naming distinction with ObjStore/Core.

File: src/cpprealm/flex_sync.hpp [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+    std::string_view object_class_name() const {
+        return m_subscription->object_class_name();
+    }
+
+    std::unique_ptr<sync::Subscription> m_subscription;
+};
+
+// A MutableSyncSubscriptionSet represents a single query that may be OR'd with other queries on the same object class to be
+// send to the server in a QUERY or IDENT message.
+struct MutableSyncSubscriptionSet {
```
Comment (jsflax) [inline src/cpprealm/flex_sync.hpp:58]: snake_case.

File: src/cpprealm/flex_sync.hpp [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+////////////////////////////////////////////////////////////////////////////
+
+#ifndef flex_sync_hpp
+#define flex_sync_hpp
+
+namespace realm {
+
+// A SyncSubscription represents a single query that may be OR'd with other queries on the same object class to be
+// send to the server in a QUERY or IDENT message.
+struct SyncSubscription {
```
Comment (jsflax) [inline src/cpprealm/flex_sync.hpp:26]: snake_case.

File: src/cpprealm/results.hpp [hunk 5]
```diff
@@ -111,6 +111,14 @@ struct results {
         return iterator(m_parent.size(), this);
     }
 
+    std::unique_ptr<T> operator[](size_t index)
```
Comment (jsflax) [inline src/cpprealm/results.hpp:114]: Should return `T` here if we can.

File: src/cpprealm/flex_sync.hpp (hunk 2 above)
Comment (leemaguire) [reply src/cpprealm/flex_sync.hpp:129]: can we not use distinguish them through namespaces instead of snake case?

**Review by jsflax — APPROVED** [review]

**Pull request merged**